File: Algorithm_347_Top K Frequent Elements.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
using namespace std;
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freqs;
        for (int i = 0; i < nums.size(); i++) freqs[nums[i]]++;
        vector<unordered_set<int> > res(nums.size() + 1);
        vector<int> toReturn;
        for (int i = 0; i < nums.size(); i++)
        {
            int pos = freqs[nums[i]];
            res[pos].insert(nums[i]);
        }
        for (int i = res.size() - 1; i >= 0; i--)
        {
            if (!res[i].empty())
            {
                unordered_set<int>::iterator iter = res[i].begin();
                while (iter != res[i].end() && k > 0)
                {
                    k--;
                    toReturn.push_back(*iter);
                    iter++;
                }
            }
        }
        return toReturn;
    }
};

// better solution
class Compare
{
public:
    bool operator()(const pair<int, int>& lhs, const pair<int, int>& rhs) {
        return lhs.second < rhs.second;
    }
};
class Solution {    
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        for(auto n : nums) {
            ++mp[n];
        }
        priority_queue<pair<int, int>, vector<pair<int, int>>, Compare> q;
        for(auto i : mp) {
            q.push(i);
        }
        vector<int> ans;
        for(int i = 0; i < k; ++i) {
            ans.push_back(q.top().first);
            q.pop();
        }
        return ans;
    }
};

File: Algorithm_347_Top K Frequent Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        buckets = [[] for _ in range(len(nums) + 1)]
        res = []
        map = {}
        for num in nums:
            count = map.get(num, 0)
            map[num] = count + 1
        for num in map:
            count = map[num]
            buckets[count].append(num)
        for i in range(len(nums), 0, -1):
            bucket = buckets[i]
            if bucket:
                for n in bucket:
                    res.append(n)
                    k -= 1
                    if k == 0:
                        return res